packages/webpack/src/buildtime/modulesData.js

```diff
@@ -0,0 +1,242 @@
+        `LavaMoatPlugin: module ${module.identifier()} has no moduleId, cannot cover it with policy.`
+      )
+      diag.rawDebug(4, { module })
+      return
+    }
+
+    // ==================================================
+    // Fixing bad modules
+
+    // Fixes the issue with assets being emitted to dist without a loader
+    // TODO: refactor to move random hardening of the build somewhere it's easier to track.
+    if (
+      isAmbientAsset(module) &&
+      module.resource.includes('node_modules') // FIXME: would be better to use canonicalName lookup and match with root
+    ) {
+      // add a warning about removing the asset
+      mainCompilationWarnings.push(
+        new WebpackError(
+          `LavaMoatPlugin: the following resource was being silently emitted to the dist directory and LavaMoat has prevented it: '${module.resource}'. If you want to add this resource, explicitly define a file-loader for it in your webpack configuration.`
+        )
+      )
+
+      // We can't use `chunkGraph.disconnectChunkAndModule` here
+      // because the require statement remains and errors out
+
+      if (module.generatorOptions) {
+        // generatorOptions was not present in testing, but types indicate it might be there
+        module.generatorOptions.emit = false
+      }
+      if (module.generator) {
+        module.generator = Object.create(module.generator, {
+          emit: {
+            value: false,
+            writable: false,
+            configurable: false,
+            enumerable: true,
+          },
+        })
+      }
+    }
+
+    // ==================================================
+    // Collecting data
+
+    // Note: module.context on an empty context module when no context information was guessable from code is going to point to the module that loads it.
+    if (isContextModule(module, moduleClass)) {
+      diag.rawDebug(3, {
+        contextModule: {
+          moduleId,
+          context: module.context,
+          // @ts-expect-error we want to see it if available
+          request: module?.options?.request,
+          // @ts-expect-error we want to see it if available
+          _identifier: module?._identifier,
+        },
+      })
+      if (!module.context) {
+        mainCompilationWarnings.push(
+          new WebpackError(
+            `LavaMoatPlugin: context module ${moduleId} has no context information. It cannot be allowed to work if it's reached at runtime.`
+          )
+        )
+      } else {
+        contextModules.push({ moduleId, context: module.context })
+      }
+    }
+    if (isIgnoredModule(module)) {
+      unenforceableModuleIds.push(moduleId)
+    } else {
+      if (isExternalModule(module, moduleClass)) {
+        externals[moduleId] = module.userRequest
+      }
+      if (isInspectableModule(module, moduleClass)) {
+        inspectable.push(module)
+      }
+
+      // typescript is complaining about the use of `resource` here, but it's actually there.
```

Comment:
```suggestion
      // typescript is complaining about the use of `resource` here, but it's actually there.
      /* c8 ignore next */
      if (!('resource' in module)) {
        // or warn or whatever
        throw Error('Expected module to have a resource')
      } 
```

This type guard (assuming it retains the `throw`) will make `module` `IdentifiedModule & {resource: unknown}` so you'd have to change yr assertion on L225 to `/** @type {string} */(module.resource)`.

...but you could also just add `resource: string` to the `IdentifiedModule` type, since you own it.  I'd still probably add the conditional & `throw` above unless I was 100% sure there's no situation in which `module.resource` is not a `string` (if it's a private API or otherwise undocumented, I wouldn't trust it to still be there after an upgrade; this will help us notice & fix the problem if it breaks).

Reply:
I'm not sure I get this. Let's talk about it tomorrow and I can fix later.